File: PlayWriter/Practical/Advance/playwright-pom-practice/pages/DashboardPage.ts
import { Locator, Page } from '@playwright/test';

export class DashboardPage {
  readonly page: Page;
  readonly section: Locator;
  readonly welcomeMessage: Locator;
  readonly logoutButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.section = page.locator('#dashboard-section');
    this.welcomeMessage = page.getByTestId('welcome-message');
    this.logoutButton = page.getByRole('button', { name: 'Logout' });
  }

  async logout() {
    await this.logoutButton.click();
  }
}

File: ManualTransition/Week2 Selenium Rebuild Properly/Practical/mini-selenium-framework-lab-practice/src/main/java/com/techlambda/selenium/miniframework/pages/DashboardPage.java
package com.techlambda.selenium.miniframework.pages;

import com.techlambda.selenium.miniframework.utils.WaitUtility;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * DashboardPage is a separate page object.
 * It keeps dashboard-related locators and dashboard actions away from test code.
 */
public class DashboardPage {

    private final WebDriver driver;
    private final WaitUtility wait;

    private final By welcomeMessage = By.id("welcome-message");
    private final By logoutButton = By.id("logout-button");
    private final By addToCartButton = By.id("add-to-cart-button");
    private final By cartMessage = By.id("cart-message");

    public DashboardPage(WebDriver driver, WaitUtility wait) {
        this.driver = driver;
        this.wait = wait;
    }

    public String getWelcomeMessage() {
        return wait.waitForVisible(welcomeMessage).getText();
    }

    public void addProductToCart() {
        wait.waitForClickable(addToCartButton).click();
    }

    public String getCartMessage() {
        return wait.waitForVisible(cartMessage).getText();
    }

    public void logout() {
        wait.waitForClickable(logoutButton).click();
    }
}
